papaya-core/src/main/java/pl/edu/agh/papaya/service/sprint/SprintService.java

```diff
@@ -94,27 +95,58 @@ Sprint createSprint(Sprint sprint) {
         return sprintRepository.findFirstByProjectIdOrderByDurationPeriodStartDesc(projectId);
     }
 
-    public Sprint closeSprint(Sprint sprint, Duration timeBurned, Duration timePlanned, LocalDateTime dateClosed) {
-        sprint.setTimeBurned(timeBurned);
-        sprint.setTimePlanned(timePlanned);
+    public Sprint closeSprint(Sprint sprint, Duration timeBurned, Duration finalTimePlanned, LocalDateTime dateClosed) {
         sprint.setDateClosed(dateClosed);
+        updateSprintStats(sprint, finalTimePlanned, timeBurned);
 
         sprintRepository.save(sprint);
 
         return sprint;
     }
 
-    public Optional<Sprint> getById(Long id) {
-        return sprintRepository.findById(id);
+    public void updateSprintStats(Sprint sprint, Duration finalTimePlanned, Duration timeBurned) {
+        SprintStats stats = sprint.getStats();
+
+        stats.setFinalTimePlanned(finalTimePlanned);
+        stats.setTimeBurned(timeBurned);
+
+        Duration totalDeclaredTime = getTotalDeclaredTimeBySprint(sprint);
+        stats.updateCoefficient(totalDeclaredTime);
+
+        stats.setAverageCoefficientCache(computeAvarageSprintCoefficient(sprint));
+    }
+
+    private Double computeAvarageSprintCoefficient(Sprint sprint) {
+        // date after start of this sprint because this one also has to be included
+        Double coefficient = sprintRepository.findAverageSprintCoefficientUpToDate(sprint.getDurationPeriod().getEnd());
+        if (coefficient == null) {
```

Comment:
Tu na pewno jest null, jak nic nie ma? Sprawdzone?

Tak na marginesie, jak się zwraca nullowalne, to pewnie i `Optionala` można tak dać i przenieść się w świat `coefficient.orElseGet(...)`

Reply:
https://stackoverflow.com/questions/52358133/spring-data-jpa-aggregate-functions-on-an-empty-resultset

optional spoko

Reply:
i chyba zapomnialem 
    @org.springframework.lang.Nullable przy tej metodzie

Reply:
https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.nullability